src/onnx/parse_convolution.cpp

```diff
@@ -42,18 +43,201 @@ struct parse_convolution : op_parser<parse_convolution>
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::half_type}}),
+            unshifted_input);
+
+        auto input_shifted_half = info.add_common_op("add", unshifted_input_half, offset_op);
+
+        return info.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::int8_type}}),
+            input_shifted_half);
+    }
+
+    static void shift_input_and_bias(const onnx_parser::node_info& info,
+                                     const instruction_ref& offset_op,
+                                     const bool has_bias,
+                                     instruction_ref& input,
+                                     instruction_ref& input_bias)
+    {
+        input = add_int8_shift(info, offset_op, input);
+        if(has_bias)
+        {
+            input_bias = add_int8_shift(info, offset_op, input_bias);
+        }
+    }
+
+    static float get_symmetric_value(const instruction_ref& input)
+    {
+        float symmetric_value = 0;
+        // adjust symmetric zero point value for uint8 types
+        if(input->get_shape().type() == migraphx::shape::uint8_type)
+        {
+            symmetric_value = 128;
+        }
+        return symmetric_value;
+    }
+
+    static instruction_ref gen_symmetric_literal(const instruction_ref& input,
+                                                 const bool is_quant_conv,
+                                                 onnx_parser::node_info& info)
+    {
+        instruction_ref ret = input;
+        if(is_quant_conv)
+        {
+            float symmetric_value = get_symmetric_value(input);
+            ret                   = info.add_literal(migraphx::literal{
+                migraphx::shape{input->get_shape().type(), {1}, {0}}, {symmetric_value}});
+        }
+
+        return ret;
+    }
+
+    static instruction_ref get_zero_point(const instruction_ref& input,
+                                          int index,
+                                          const bool is_quant_conv,
+                                          onnx_parser::node_info& info,
+                                          const std::vector<instruction_ref>& args)
+    {
+        instruction_ref ret = input;
+        if(args.size() > index)
+        {
+            // Check for type mismatch on parse
+            if(input->get_shape().type() != args[index]->get_shape().type())
+                MIGRAPHX_THROW("PARSE:Conv Data and Data Zero Point must have same type");
+
+            ret = args[index];
+            if(is_symmetric_zero_point(ret))
+            {
+                std::cout << "Hit symmetric input" << std::endl;
+                ret = gen_symmetric_literal(ret, is_quant_conv, info);
+            }
+            ret->debug_print();
+        }
+        else
+        {
+            ret = gen_symmetric_literal(ret, is_quant_conv, info);
+        }
+
+        return ret;
+    }
+
+    static bool is_symmetric_zero_point(instruction_ref zp)
```

Comment:
why should one care if it is symmetrical or not.. a simple add operator works with or without symmetry, Thanks..

Reply:
This is relevant as we're avoiding overflow + reducing terms created when we parse in a Convolution and this leads to a few simplifications + helps accuracy .

In this operator we're performing convolution of two sums which simplifies too

`Conv(X - A, Y - B) =  Conv(X,Y) - Conv(X, B) - Conv(Y,A) + Conv(A,B)`

Convolution becomes a a set of matrix multiplication and additions of terms.

For the int8 case, the multiply in the convolution in Int8, MatMulInteger(X,Y) between terms must not overflow but the accumulation can as each MatMulInteger op accumulates in int32. Expanding the terms out ensures we wont roll over/have accuracy issues when performing a bias before convolution.

`MatMulInteger(X - A, Y - B) =  MatMulInteger(X,Y) - MatMulinteger(X, B) - MatMulinteger(Y,A) + MatMulInteger(A,B)`

If a zero point is symmetric (128 in uint8, and 0 in int8) one of the four terms before the convolution is performed becomes zero, thus we don't create two of the terms in the expansion.

That's why in the code we're checking for symmetric with respect to datatype and I've expanded it from a regular inline if to function calls to get and generate symmetric zero points. Should we start adding other quantized types, we can just add the data type to the list with is specific zero point.

Reply:
Pardon me, what I meant was, there is an 'add' step (of a constant value 128) for `bias` and `input` (or whatever..) in `add_int8_shift()`. Then we have a 'subract' step in `handle_quant_bias()`. Couldn't there be just one 'subract' of a value = zp - 128 ? And avoid the 'add' step for a constant '128'.

Reply:
BTW, it can be a simple 'subtract' as I mention above.. Or just an 'add' if the quantities were reversed. But the point is, do we need both: a 'subtract' with the ZP + an 'add' with '128'?

Reply:
OH I see what you're saying. There's two things going on here that are not related

1. The add_int8_shift() shifts the input (data + bias value) so that it operates in int8 if its uint8 if that uint8 is a symmetric zero point (128) thus we subtract and do the add of a negative -128 value to data and zero point.

2. the handle_quant_bias() applies the bias  shifted convolution step (after the datatype is shifted from uint8 to int8) with the post shifted biases since the result is expanded as I explained above. To expand things it requires us to have identical datatypes as each Conv resolves to a GEMM, and mismatching GEMMs (uint8 and int8) aren't permitted. The result of each of the expanded convolutions is int32 and thus we dont need to worry as much about overflow of the result when we add the resulting convolutions

The subtraction isn't for the bias here, its to subtract Convolutions between biases and zero points which would be negative when we expand out Conv(X - a, Y - b)

Reply:
As per our discussion, this is a const-folding topic, so not an issue for me that there are effectively two 'add' operation, instead of just one.